lib/clockwise_import.rb

```diff
@@ -0,0 +1,211 @@
+      end
+    end
+  end
+
+  def find_most_recent_entry_date(project)
+    date = project.entries.order(:date).last.date
+    DateTime.parse date.to_s
+  end
+
+end
+
+class EntryBuilder
+  def initialize(row)
+    @row = row
+  end
+
+  def create!
+    client = build_client
+    project = build_project(client)
+    category = build_category
+    user = build_user
+    Entry.create!(
+      project: project,
+      category: category,
+      user: user,
+      hours: @row.hours,
+      date: @row.date,
+      description: @row.description,
+      billed: @row.archive
+    )
+  end
+
+  private
+
+  def build_client
+    Client.where('lower(name) = ?', @row.client.downcase).
+      first_or_create(name: @row.client)
+  end
+
+  def build_project(client)
+    Project.where('lower(name) = ? or slug = ?',
+                  @row.project.downcase,
+                  @row.project.parameterize).
+      first_or_create(name: @row.project,
+                      billable: true,
+                      client: client,
+                      archived: @row.archive)
+  end
+
+  def build_category
+    Category.where('lower(name) = ?', @row.category.downcase).
+      first_or_create(name: @row.category)
+  end
+
+  def build_user
+    names = @row.user.split(' ')
+    first_name = names.first
+    last_name = names.last
+    User.where('lower(first_name) = ? and lower(last_name) = ?',
+               first_name.downcase, last_name.downcase).
+                first_or_create(first_name: first_name,
+                                last_name: last_name,
+                                email: build_email_from_name(@row.user),
+                                password: SecureRandom.hex,
+                                confirmed_at: DateTime.now)
+  end
+
+  def build_email_from_name(name)
+    names = name.split(' ')
+    "#{names[0][0]}.#{names[1]}@defacto.nl".downcase
+  end
+end
+
+class ImportRow
+  attr_reader :archive,
+              :user,
+              :client,
+              :project,
+              :category,
+              :date,
+              :hours,
+              :description
+
+  def initialize(row)
+    @archive = row[0] == 'J'
+    @user = row[1]
+    @client = row[2]
+
+    if row[3].to_s.empty? || row[2].downcase == 'defacto'
+      @project = row[2]
+    else
+      @project = row[3]
+    end
+
+    if row[4].to_s.empty?
+      @category = 'Clockwise Import'
+    else
+      @category = row[4]
+    end
+
+    @date = parse_date(row[7])
+    @hours = parse_full_hours(row[8])
+    @description = String(row[9]).truncate(255)
+  end
+
+  private
+
+  def parse_date(date_string)
+    comps = date_string.split('/')
+    year = Integer('20' + comps[2].last(2))
+    day = Integer(comps[1])
+    month = Integer(comps[0])
+    DateTime.new(year, month, day)
+  end
+
+  def parse_full_hours(hours_string)
+    String(hours_string).gsub(',','.').to_f.round
```

Comment:
Space missing after comma.<br>Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.